src/xkbcomp/types.c

```diff
@@ -180,6 +180,11 @@ MergeIncludedKeyTypes(KeyTypesInfo *into, KeyTypesInfo *from,
 {
     if (from->errorCount > 0) {
         into->errorCount += from->errorCount;
+        // Free unused types
```

Comment:
If I'm reading correctly, the "contract" of `MergeIncludedKeyTypes` is that it steals/empties out `from` from within, but leaves it in tact; the caller still needs to clear `from`.  And `HandleIncludeKeyTypes` does call `ClearKeyTypesInfo` after each `MergeIncludedKeyTypes` call.

So I suppose the question is, shouldn't `ClearKeyTypesInfo` take care of freeing `types`?

Reply:
That was my first thought, but then I ran into double-free issues.

Reply:
Yep but I think that's what we should fix. I think we need this:

```diff
diff --git a/src/xkbcomp/types.c b/src/xkbcomp/types.c
index 34e1028..9f499ad 100644
--- a/src/xkbcomp/types.c
+++ b/src/xkbcomp/types.c
@@ -201,6 +201,7 @@ MergeIncludedKeyTypes(KeyTypesInfo *into, KeyTypesInfo *from,
             if (!AddKeyType(into, type, false))
                 into->errorCount++;
         }
+        darray_init(from->types);
     }
 }
```

Reply:
So I re-analyze the issue and I think I get it right this time. I left comments in the code to explain it, as it was difficult to detect for me.

Reply:
And I fixed another leak in `HandleKeyTypeDef`.